seqr/utils/vcf_utils.py

```diff
@@ -0,0 +1,91 @@
+import re
+
+from collections import defaultdict
+
+from seqr.utils.middleware import ErrorsWarningsException
+from seqr.utils.file_utils import file_iter
+
+BLOCK_SIZE = 65536
+
+EXPECTED_META_FIELDS ={
+    'INFO': {
+        'AC': {'Number': 'A', 'Type': 'Integer'},
+        'AN': {'Number': '1', 'Type': 'Integer'},
+        'AF': {'Number': 'A', 'Type': 'Float'}
+    },
+    'FORMAT': {
+        'AD': {'Number': '.', 'Type': 'Integer'},
+        'DP': {'Number': '1', 'Type': 'Integer'},
+        'GQ': {'Number': '1', 'Type': 'Integer'},
+        'GT': {'Number': '1', 'Type': 'String'}
+    }
+}
+
+REQUIRED_HEADERS = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT']
+
+
+def _validate_vcf_header(header):
+    missing_headers = [h for h in REQUIRED_HEADERS if h not in header ]
+    if missing_headers:
+        missing_fields = ','.join(missing_headers)
+        raise ErrorsWarningsException([f'Required VCF header fields {missing_fields} are missing.'], [])
+
+
+def _validate_vcf_meta(meta):
+    for field, sub_field_meta in EXPECTED_META_FIELDS.items():
+        for sub_field, metadata in sub_field_meta.items():
+            for key, value in metadata.items():
+                if meta.get(field, {}).get(sub_field, {}).get(key) != value:
+                    error = f'VCF header field {field}.{sub_field} and meta information {key}={value} is expected.'
+                    raise ErrorsWarningsException([error], [])
+
+
+def _validate_vcf_data_row(first_line):
+    # Has required INFO fields?
+    row_data = first_line.split('\t')
+    info_keys = [k.split('=')[0] for k in row_data[REQUIRED_HEADERS.index('INFO')].split(';')]
+    missing_info_fields = [f for f in EXPECTED_META_FIELDS['INFO'].keys() if f not in info_keys]
+    if missing_info_fields:
+        missing_fields = ','.join(missing_info_fields)
+        raise ErrorsWarningsException([f'Missing INFO field(s) {missing_fields} in the row data'], [])
+
+    # Has required FORMAT fields?
+    format_keys = [k for k in row_data[REQUIRED_HEADERS.index('FORMAT')].split(':')]
+    missing_format_fields = [f for f in EXPECTED_META_FIELDS['FORMAT'].keys() if f not in format_keys]
+    if missing_format_fields:
+        missing_fields = ','.join(missing_format_fields)
+        raise ErrorsWarningsException([f'Missing FORMAT field(s) {missing_fields} in the row data'], [])
+
+
+def _get_vcf_meta_data(line):
+    r = re.search(r'##(.*?)=<(.*)>$', line)
+    if r:
+        field, value = r.groups()
```

Comment:
it is better to used named groups to be explicit about what you are parsing from where:
using `r'##(?P<field>.*?)=<(?P<value>.*)>$'` if you call `r.groupdict()` you will get a dictionary keys `field` and `value`

I also think there is a better way to write this regex so that it can capture the subfields we care about directly. I think we only want to validate on ID and Type, so using `r'##(?P<field>.*?)=<ID=(?P<id>[^,]*).*Type=(?P<type>[^,]*).*>$'` will explicitly capture those subfields

Reply:
It is great to use `groupdict`. But I'm unsure whether it works for picking up the `ID` and `Type` fields because their order is not determined.

Reply:
But the `ID` must be the first one.

Reply:
if ID must be first than the regex I listed will work regardless of what other fields are there. If thats not actually true of the metadata then we can rethink this, but I think it would still be better to write an explicit regex for ID and Type than parsing out all the subfields